taskvine/src/worker/vine_worker.c

```diff
@@ -1291,6 +1306,87 @@ static int enforce_worker_promises(struct link *manager)
 	return 1;
 }
 
+/*
+Given a freshly started process, wait for it to initialize and send
+back the library startup message with JSON containing the name of
+the library, which should match the task's provides_library label.
+*/
+
+static int check_library_startup(struct vine_process *p)
+{
+	char buffer_len[VINE_LINE_MAX];
+	int length = 0;
+
+	/* Read a line that gives the length of the response message. */
+	if (!link_readline(p->library_read_link, buffer_len, VINE_LINE_MAX, LINK_NOWAIT)) {
+		return 0;
+	}
+	sscanf(buffer_len, "%d", &length);
+
+	/* Now read that length of message and null-terminate it. */
+	char buffer[length + 1];
+	if (link_read(p->library_read_link, buffer, length, LINK_NOWAIT) <= 0) {
+		return 0;
+	}
+	buffer[length] = 0;
+
+	/* Check that the response is JX and contains the expected name. */
+	struct jx *response = jx_parse_string(buffer);
+
+	const char *name = jx_lookup_string(response, "name");
+
+	int ok = 0;
+	if (!strcmp(name, p->task->provides_library)) {
+		ok = 1;
+	}
+	if (response) {
+		jx_delete(response);
+	}
+	return ok;
+}
+
+/* Check whether all known libraries are ready to execute functions.
+ * A library starts up and tells the vine_worker it's ready by reporting
+ * back its library name. */
+static void check_libraries_ready()
+{
+	uint64_t library_task_id;
+	struct vine_process *library_process;
+
+	struct link_info library_link_info;
+	library_link_info.events = LINK_READ;
+	library_link_info.revents = 0;
+
+	/* Loop through all known libraries and check if they are ready. */
+	ITABLE_ITERATE(libraries_table, library_task_id, library_process)
+	{
+		if (!library_process->library_ready) {
+			/* Skip library processes that haven't been started. */
+			if (!library_process->library_read_link) {
+				continue;
+			}
+			library_link_info.link = library_process->library_read_link;
```

Comment:
`link_sleep` gives a simpler way to check if a single link is readable or writeable.

Reply:
How do you set the stoptime parameter for `link_sleep`? time(0)+500ms?

Reply:
Look like `link_sleep` takes a `time_t` but `link_usleep` takes relative milliseconds:
https://cctools.readthedocs.io/en/latest/api/html/link_8h.html#a88872a9b3d0490d2f690e34486786ac5

Reply:
In terms of performance I think link_poll is better right, since it doesn't block or depends on timeouts?

Reply:
They all end up calling `poll` one way or the other, just trying to show a way that is more direct...